standalone/elf.c

```diff
@@ -57,61 +56,94 @@ gen_elf_segment(uint8_t **code, uintptr_t target, void *src, size_t len,
-  mbi->flags |=  MBI_FLAG_CMDLINE;
-
   // check elf header
-  struct eh *elf = (struct eh *) m->mod_start;
+  struct eh *elf = (struct eh *) binary;
   assert(memcmp(elf->e_ident, ELFMAG, SELFMAG) == 0, "ELF header incorrect");
 
-  uint8_t *code = (uint8_t *)0x7C00;
-
-#define LOADER(EH, PH) {                                                \
-    struct EH *elfc = (struct EH *)elf;                                               \
+#define ITER(EH, PH) {                                                  \
+    struct EH *elfc = (struct EH *)elf;                                 \
     assert(elfc->e_type==2 && ((elfc->e_machine == EM_386) || (elfc->e_machine == EM_X86_64)) && elfc->e_version==1, "ELF type incorrect"); \
     assert(sizeof(struct PH) <= elfc->e_phentsize, "e_phentsize too small"); \
                                                                         \
     for (unsigned i = 0; i < elfc->e_phnum; i++) {                      \
-      struct PH *ph = (struct PH *)(uintptr_t)(m->mod_start + elfc->e_phoff+ i*elfc->e_phentsize); \
-      if (ph->p_type != 1)                                              \
-        continue;                                                       \
-      gen_elf_segment(&code, ph->p_paddr, (void *)(uintptr_t)(m->mod_start+ph->p_offset), ph->p_filesz, \
-                      ph->p_memsz - ph->p_filesz);                      \
-    }                                                                   \
-                                                                        \
-    gen_mov(&code, EAX, 0x2BADB002);                                    \
-    gen_mov(&code, EDX, elfc->e_entry);                                 \
+      struct PH *ph = (struct PH *)(uintptr_t)(binary + elfc->e_phoff+ i*elfc->e_phentsize); \
+      struct ph64 const p_tmp = {      \
+        .p_type   = ph->p_type,        \
+        .p_flags  = ph->p_flags,       \
+        .p_offset = ph->p_offset,      \
+        .p_vaddr  = ph->p_vaddr,       \
+        .p_paddr  = ph->p_paddr,       \
+        .p_filesz = ph->p_filesz,      \
+        .p_memsz  = ph->p_memsz,       \
+        .p_align  = ph->p_align        \
+      };                               \
+      int e = fn(&p_tmp, binary, var); \
+      if (e)                           \
+        return e;                      \
+    }                                  \
+  }
+
+  switch (elf->e_ident[EI_CLASS]) {
+  case ELFCLASS32:
+    ITER(eh, ph);
+    break;
+  case ELFCLASS64:
+    ITER(eh64, ph64);
+    break;
+  default:
+    assert(false, "Invalid ELF class");
   }
+  return 0;
+}
+
+static int
+elf_phdr(struct ph64 const *ph, uint32_t const binary, uint8_t ** code)
+{
+  if (ph->p_type != ELF_PT_LOAD)
+    return 0;
+
+  gen_elf_segment(code, ph->p_paddr, (void *)(uintptr_t)(binary+ph->p_offset),
+                  ph->p_filesz, ph->p_memsz - ph->p_filesz);
+
+  return 0;
+}
+
+int
+load_elf(void const * mbi, uint32_t const binary, uint32_t const magic,
```

Comment:
If `binary` is always holding a pointer, it would be nice if it's either `void *`, `char *` or `uintptr_t`.